issue_95: Convert CounterOne.jsx to Function Component

soundarya-mv opened the pull request:
Convert class component to function component
Moving static data to database.json

## Rationale

Convert class component to function component
Moving static data to database.json

## Advice for Reviewers & Testing Notes

Home page counter component needs to be tested.

## Screenshots:

- Add a screenshot or two of your changes, where applicable.

## Issue Resolved

*Note - This allows the issue to be automatically linked to the PR once the PR is submitted for review*
- Fixes #95

soundarya-mv committed:
issue_95: Convert CounterOne.jsx to Function Component

Convert class component to function component
Moving static data to database.json

davidtaing commented on public/database.json:
```diff
@@ -258,5 +258,31 @@
       "avatar": "WD_Ella.png"
     }
-  ]
+  ],
+    "counterData":[
+      {
+        "countNum": 3000,
+        "countTitle": "Meetup Members"
+      },
+      {
+        "countNum": 1000,
+        "countTitle": "Social Media Followers"
+      },
+      {
+        "countNum": 200,
+        "countTitle": "Meetups"
+      },
+      {
+        "countNum": 100,
+        "countTitle": "Discord Members"
+      },
+      {
+        "countNum": 20,
+        "countTitle": "Projects"
+      },
+      {
+        "countNum": 15,
+        "countTitle": "Courses"
+      }
+    ]
```
**Nitpick:** 
For this I would have preferred for this to be done as a separate commit. In my opinion, we can group the changes as:
1. Migrate `CounterOne` component to functional component
2. Refactor `counterData` to `database.json` file

Minor and (probably a non-issue) since this is my personal opinion.

davidtaing commented on src/elements/counters/CounterOne.jsx:
```diff
@@ -1,69 +1,41 @@
+import React, { useState, Fragment } from "react"
 import CountUp from "react-countup"
 import VisibilitySensor from "react-visibility-sensor"
+import database from "../../../public/database.json";
```
**Optional:**
I just thought about this (and sorry for not raising this up in the earlier PRs). Perhaps we could destructure `counterData` here?

Feel free to leave this as it is or destructure. I'll leave that one up to you.

soundarya-mv replied:
@davidtaing  We have a folder named counter with 2 files as counterOne and counterTwo. I thought it was more structured when it is grouped that way. Please let me know, if we can remove the folder and have the 2 files in the elements folder directly.

davidtaing replied:
For context, this project was started from a paid React template. That's why we have multiple versions of the components in the codebase.

Personally, I've got no problems with moving to a more simplified structure (and even deleting the extra versions). But I'll defer the decision to our benevolent director. :smile: 

@engramar, what do you think? Should we keep the file structure untouched or go ahead and move the files into the `elements` folder?

davidtaing commented on src/elements/counters/CounterOne.jsx:
```diff
@@ -1,69 +1,41 @@
-      this.setState({ didViewCountUp: true })
+      setState({ didViewCountUp: true })
     }
   }
```
Using `state` as the name for your state hook isn't great, because it doesn't offer any contextual information.

One of the questions that I like to ask myself is "Will I understand this in 6 months?"

So think about future you. Will you be able to understand what `state` does in 6 months?

soundarya-mv replied:
Updated

davidtaing replied:
It's kinda my bad, but I should have mentioned that we don't need to use an object and instead directly use a boolean value.

Optional item for later. I'll merge this PR in for now.

Thanks for the good work.

soundarya-mv replied:
yes @davidtaing , even I thought of making that as a boolean data. But thought it was defined as an object intentionally for some reason expecting more data might be added to the object. So maintained it as the same.

davidtaing commented on src/elements/counters/CounterOne.jsx:
```diff
@@ -1,69 +1,41 @@
-    )
-  }
+  return (
+    <Fragment>
```
**Optional:**
If we go with the following syntax for Fragment, we can drop the import statement at the top of the file.

```JSX
<>
  // ...rest of component
</>
```

soundarya-mv replied:
Change done

davidtaing reviewed (changes requested):
Thanks for your PR @soundarya-mv, I've left four comments below. Please go through them when you get a chance.

I've marked two as `optional`, so you can either add additional changes or leave them as is. 

And I've marked another as a `nitpick`, due to it being a personal preference. No action required for the `nitpick` comments.

soundarya-mv committed:
issue_95: Convert CounterOne.jsx to Function Component

1) Updating the name of the state hook to a more readable one.
2) Using <> instead of <fragment> tag

soundarya-mv committed:
Merge branch 'main' into issue_95
